placement/algo/algo.go

```diff
@@ -0,0 +1,470 @@
+		return nil, err
+	}
+	return ph.generatePlacement(), nil
+}
+
+func (a rackAwarePlacementAlgorithm) AddReplica(ps placement.Snapshot) (placement.Snapshot, error) {
+	ph := newReplicaPlacementHelper(ps, ps.Replicas()+1)
+	if err := ph.placeShards(ph.uniqueShards, nil); err != nil {
+		return nil, err
+	}
+	return ph.generatePlacement(), nil
+}
+
+func (a rackAwarePlacementAlgorithm) RemoveHost(ps placement.Snapshot, leavingHost placement.Host) (placement.Snapshot, error) {
+	if a.isHostAbsent(ps, leavingHost) {
+		return nil, errHostAbsent
+	}
+	ph, leavingHostShards := newRemoveHostPlacementHelper(ps, leavingHost)
+	// place the shards from the leaving host to the rest of the cluster
+	if err := ph.placeShards(leavingHostShards.Shards(), leavingHostShards); err != nil {
+		return nil, err
+	}
+	return ph.generatePlacement(), nil
+}
+
+func (a rackAwarePlacementAlgorithm) AddHost(ps placement.Snapshot, addingHost placement.Host) (placement.Snapshot, error) {
+	ph, addingHostShards := newAddHostPlacementHelper(ps, addingHost)
+	targetLoad := ph.hostHeap.getTargetLoadForHost(addingHostShards.hostAddress())
+	// try to steal shards from the most loaded hosts until the adding host reaches target load
+outer:
+	for len(addingHostShards.shardsSet) < targetLoad {
+		if ph.hostHeap.Len() == 0 {
+			return nil, errCouldNotReachTargetLoad
+		}
+		tryHost := heap.Pop(ph.hostHeap).(*hostShards)
+		for shard := range tryHost.shardsSet {
+			if ph.canAssignHost(shard, tryHost, addingHostShards) {
+				ph.assignShardToHost(shard, addingHostShards)
+				ph.removeShardFromHost(shard, tryHost)
+				heap.Push(ph.hostHeap, tryHost)
+				continue outer
+			}
+		}
+	}
+
+	return ph.generatePlacement(), nil
+}
+
+func (a rackAwarePlacementAlgorithm) ReplaceHost(ps placement.Snapshot, leavingHost, addingHost placement.Host) (placement.Snapshot, error) {
+	if a.isHostAbsent(ps, leavingHost) {
+		return nil, errHostAbsent
+	}
+	addingHostShards := newEmptyHostShardsFromHost(addingHost)
+	ph, leavingHostShards := newRemoveHostPlacementHelper(ps, leavingHost)
+
+	var shardsUnassigned []uint32
+	// move shards from leaving host to adding host
+	for _, shard := range leavingHostShards.Shards() {
+		if ph.canAssignHost(shard, leavingHostShards, addingHostShards) {
+			ph.removeShardFromHost(shard, leavingHostShards)
+			addingHostShards.addShard(shard)
+		} else {
+			shardsUnassigned = append(shardsUnassigned, shard)
+		}
+	}
+	// if there are shards that can not be moved to adding host
+	// distribute them to the cluster
+	if err := ph.placeShards(shardsUnassigned, leavingHostShards); err != nil {
+		return nil, err
+	}
+
+	cl := ph.generatePlacement()
+	// add the adding host to the cluster and bring its load up to target load
+	cl.hostShards = append(cl.hostShards, addingHostShards)
+	return a.AddHost(cl, addingHost)
+}
+
+func (a rackAwarePlacementAlgorithm) isHostAbsent(ps placement.Snapshot, h placement.Host) bool {
+	for _, hs := range ps.HostShards() {
```

Comment:
Perhaps it'd be better to define a `HasHost` method on the `Snapshot` interface, which internally uses a map to look up a host rather than looping through a list.